### PR TITLE
Nc games 05

### PR DESCRIPTION
**Pull request by LexJoss**
Refactor of task 5 completed
Includes completed task 6
refactor of sql in "get reviews" to mean that reviews with no comments no longer so "1", and correctly show "0"

Please excuse that name being 5.

### CODE REVIEW

File: __tests__/models.test.js [hunk 1]
```diff
@@ -45,6 +49,7 @@ describe("2nd endpoint, reviews, can return status 200", () => {
         .then(response => {
             const reviews = response.body
             reviews.forEach((review) => {
+                console.log(review)
```
Comment (two-triangles) [inline __tests__/models.test.js]: remove logs that are not needed.

File: __tests__/models.test.js [hunk 2]
```diff
@@ -82,6 +84,77 @@ describe("2nd endpoint, reviews, can return status 200", () => {
...
+    })
+    test("The response body only contains one element, which has the correct ID", () => {
+        return request(app).get('/api/reviews/1').expect(200)
+        .then(response => {
+            const review = response.body
+            expect(review.length).toBe(1)
+            expect(review[0].review_id).toBe(1)
+        })
+    }) 
+    test("Will respond with an error message if an unavailable ID is given", () => {
```
Comment (two-triangles) [inline __tests__/models.test.js:113]: We need to add a 400 test for an id that is an invalid type - /api/reviews/cheese for example

File: __tests__/models.test.js [hunk 3]
```diff
@@ -82,6 +84,77 @@ describe("2nd endpoint, reviews, can return status 200", () => {
...
+})
+describe("4th endpoint, comments by review id", () => {
+    test("4th endpoint responds with a status 200", () => {
+        return request(app).get('/api/reviews/2/comments').expect(200)
+    })
+    test("The return object has the correct keys for a comment", () => {
+        return request(app).get('/api/reviews/2/comments').expect(200)
+        .then (response => {
+            const comments = response.body
+            comments.forEach((comment) => {
```
Comment (two-triangles) [inline __tests__/models.test.js:136]: please add an assertion for the length before running this forEach() - this is important as if the returned rows were empty, the forEach would not be executed.

File: __tests__/models.test.js [hunk 4]
```diff
@@ -82,6 +84,77 @@ describe("2nd endpoint, reviews, can return status 200", () => {
...
+            expect(comment).toHaveProperty('author', expect.any(String))
+            expect(comment).toHaveProperty('body', expect.any(String))
+            expect(comment).toHaveProperty('review_id', expect.any(Number))
+        })
+    })
+    })
+    test('The return object resolves with the correct review_id per comment', () => {
+        return request(app).get('/api/reviews/2/comments').expect(200)
+        .then (response => {
+            const comments = response.body
```
Comment (two-triangles) [inline __tests__/models.test.js:149]: see comment above

File: __tests__/models.test.js [hunk 5]
```diff
@@ -82,6 +84,77 @@ describe("2nd endpoint, reviews, can return status 200", () => {
...
+    })
+    test("Will respond with an error message if an unavailable ID is given", () => {
+        return request(app).get('/api/reviews/1/comments').expect(404)
+        .then((response) => {
+            expect(response.body.msg).toBe("Not Found")
+        })
+    })
+})
+
+
```
Comment (two-triangles) [inline __tests__/models.test.js]: error needed for invalid id_type - 400

File: db/models.js [hunk 6]
```diff
@@ -20,8 +20,40 @@ const fetchReviews = () => {
...
+        if (rows.length === 0) {
+            return Promise.reject({status: 404, msg: "Not Found"})
+        } else {
+            return rows}
+    })
+}
+        
 
+const fetchCommentsByRid = (review_id) => {
+    console.log(typeof review_id.review_id)
```
Comment (two-triangles) [inline db/models.js]: remove console.logs that are not needed

File: db/models.js [hunk 7]
```diff
@@ -1,27 +1,64 @@
...
-        return db.query(sqlString)
-        .then (({rows}) => rows)
-    }
+
+    return db.query(sqlString)
+        .then(({ rows }) => rows)
+}
+
+const fetchReviewByID = (review_id) => {
+    if (isNaN(review_id)) {return Promise.reject({status : 400, msg : "Not Found"})}
```
Comment (two-triangles) [inline db/models.js]: This is sufficient for how we want our server to behave but it is worth pointing out that postgres would throw an error here regardless as the review_id has to be an INT. We could use that throw error and pass it on to some error handling middleware.

File: db/models.js [hunk 8]
```diff
@@ -1,27 +1,64 @@
...
+    const sqlString = `SELECT * FROM reviews
+    WHERE reviews.review_id = $1;`
 
 
+    return db.query(sqlString, [review_id])
+        .then(({ rows }) => {
+            if (rows.length === 0) {
+                return Promise.reject({ status: 400, msg: "Not Found" })
```
Comment (two-triangles) [inline db/models.js]: This should be 404

File: db/models.js [hunk 9]
```diff
@@ -1,27 +1,64 @@
...
+            } else {
+                return rows
+            }
+        })
+}
+
+
+const fetchCommentsByRid = (review_id) => {
+    const values = [review_id.review_id]
+    if (isNaN(values)) {return Promise.reject({status : 400, msg : "Not Found"})}
```
Comment (two-triangles) [inline db/models.js]: same as comment above regarding postgres throwing the error

File: db/models.js [hunk 10]
```diff
@@ -1,27 +1,64 @@
...
+    const sqlString = `SELECT * FROM comments
+    WHERE comments.review_id = $1;`
+
+    return fetchReviewByID(review_id.review_id)
+        .then(() => { return db.query(sqlString, values) })
+        .then(({ rows }) => {
+            if (rows.length === 0) {
```
Comment (two-triangles) [inline db/models.js:49]: Is this check necessary? You fetchReviewById function is returning a 400(should be 404) if the rows is empty anyway? So do we need to check this again?

File: db/models.js [hunk 11]
```diff
@@ -1,27 +1,64 @@
...
-    .then (({rows}) => rows)
+        .then(({ rows }) => rows)
 }
 
 
 const fetchReviews = () => {
-    
-    let sqlString = `SELECT reviews.*, count(*) AS comment_count 
+
+    let sqlString = `SELECT reviews.*, count(comments.review_id) AS comment_count
```
Comment (two-triangles) [inline db/models.js:13]: Just be aware that your comment_count is coming back as a string. This is something we will have to accommodate for in FE.

File: __tests__/models.test.js [hunk 12]
```diff
@@ -82,6 +83,104 @@ describe("2nd endpoint, reviews, can return status 200", () => {
...
+            expect(comment).toHaveProperty('body', expect.any(String))
+            expect(comment).toHaveProperty('review_id', expect.any(Number))
+        })
+    })
+    })
+    test('The return object resolves with the correct review_id per comment', () => {
+        return request(app).get('/api/reviews/2/comments').expect(200)
+        .then (response => {
+            const comments = response.body
+            comments.forEach((comment) =>{
```
Comment (MrMcSnail) [inline __tests__/models.test.js:151]: Don't forget a length assertion here

File: __tests__/models.test.js [hunk 13]
```diff
@@ -82,6 +83,104 @@ describe("2nd endpoint, reviews, can return status 200", () => {
...
+            expect(response.body.msg).toBe("Not Found")
+        })
+    })
+    test("Will respond with an error message if an invalid parameter is given", () => {
+        return request(app).get('/api/reviews/cheese').expect(400)
+        .then((response) => {
+            expect(response.body.msg).toBe("Bad Request")
+        })
+    })
+    test("Will respond with an empty array if the parent review has no comments", () => {
```
Comment (MrMcSnail) [inline __tests__/models.test.js:168]: good stuff.

**Review by MrMcSnail — APPROVED** [review]

**Pull request merged**